src/libsyntax/ext/expand.rs

```diff
@@ -640,6 +640,9 @@ pub fn expand_item_mac(it: P<ast::Item>, fld: &mut MacroExpander)
             // need to be marked. Not that it could be marked anyway.
             // create issue to recommend refactoring here?
             fld.cx.syntax_env.insert(intern(name[]), ext);
+            // Keep track of this macro definition both within the crate
+            // context and (if applicable) in its exports.
+            fld.cx.macros.push(it.clone());
```

Comment:
Yeah I think something like that is the way to go. I'd probably avoid `ast::Crate` and have some other kind of side-structure though as we should probably keep the AST limited to being an AST :)